app/services/fact_importer.rb

```diff
@@ -40,26 +40,62 @@ def initialize(host, facts = {})
   # expect a facts hash
   def import!
     ActiveRecord::Base.transaction do
+      ensure_fact_names
       delete_removed_facts
       update_facts
       add_new_facts
     end
 
-    if @error
-      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
-      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s"), @host.name)
-    end
+    report_error(@error) if @error
     logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
   end
 
+  def report_error(error)
+    Foreman::Logging.exception("Error during fact import for #{@host.name}", error)
+    raise ::Foreman::WrappedException.new(error, N_("Import of facts failed for host %s"), @host.name)
+  end
+
   # to be defined in children
   def fact_name_class
     raise NotImplementedError
   end
 
   private
 
-  attr_reader :host, :facts
+  attr_reader :host, :facts, :fact_names
+
+  def ensure_fact_names
+    initialize_fact_names
+
+    missing_keys = facts.keys - fact_names.keys
+    add_missing_fact_names(missing_keys)
+  end
+
+  def add_missing_fact_names(missing_keys)
+    missing_keys.sort.each do |fact_name|
+      # create a new record and make sure it could be saved.
+      name_record = fact_name_class.new(fact_name_attributes(fact_name))
+      name_record.validate!
+
+      # if the record could not be saved, then it should already be in the database
+      # throw an exception if it isn't
+      name_record = fact_name_class.find_by!(name: fact_name) unless name_record.save
+
+      # make the new record available immediately for other fact_name_attributes calls
+      @fact_names[fact_name] = name_record
+    end
+  end
+
+  def initialize_fact_names
+    name_records = fact_name_class.where(:name => facts.keys, :type => fact_name_class.to_s).to_a
+    @fact_names = Hash[name_records.map{ |record| [record.name, record]}]
```

Comment:
You can use `fact_name_class.where(:name => facts.keys, :type => fact_name_class.to_s).group_by(&:name)` to reduce object allocations and clean up the code